src/main/java/jcifs/netbios/NameServiceClientImpl.java

```diff
@@ -0,0 +1,1072 @@
+                return response.addrEntry;
+            }
+        }
+        while ( --n > 0 && request.isBroadcast );
+
+        throw new UnknownHostException(name.name);
+    }
+
+
+    NbtAddress getByName ( Name name, InetAddress addr ) throws UnknownHostException {
+        NameQueryRequest request = new NameQueryRequest(this.transportContext.getConfig(), name);
+        NameQueryResponse response = new NameQueryResponse(this.transportContext.getConfig());
+
+        if ( addr != null ) { /*
+                               * UniAddress calls always use this
+                               * because it specifies addr
+                               */
+            request.addr = addr; /* if addr ends with 255 flag it bcast */
+            request.isBroadcast = ( addr.getAddress()[ 3 ] == (byte) 0xFF );
+
+            int n = this.transportContext.getConfig().getNetbiosRetryCount();
+            do {
+                try {
+                    send(request, response, this.transportContext.getConfig().getNetbiosRetryTimeout());
+                }
+                catch ( InterruptedIOException ioe ) {
+                    if ( log.isTraceEnabled() ) {
+                        log.trace("Timeout waiting for response " + name.name, ioe);
+                    }
+                    throw new UnknownHostException(name.name);
+                }
+                catch ( IOException ioe ) {
+                    log.info("Failed to send nameservice request for " + name.name, ioe);
+                    throw new UnknownHostException(name.name);
+                }
+
+                if ( response.received && response.resultCode == 0 ) {
+                    int last = response.addrEntry.length - 1;
+                    response.addrEntry[ last ].hostName.srcHashCode = addr.hashCode();
+                    return response.addrEntry[ last ];
+                }
+            }
+            while ( --n > 0 && request.isBroadcast );
+
+            throw new UnknownHostException(name.name);
+        }
+
+        /*
+         * If a target address to query was not specified explicitly
+         * with the addr parameter we fall into this resolveOrder routine.
+         */
+
+        for ( ResolverType resolverType : this.resolveOrder ) {
+            try {
+                switch ( resolverType ) {
+                case RESOLVER_LMHOSTS:
+                    NbtAddress ans = this.lmhosts.getByName(name, this.transportContext);
+                    if ( ans != null ) {
+                        ans.hostName.srcHashCode = 0; // just has to be different
+                                                      // from other methods
+                        return ans;
+                    }
+                    break;
+                case RESOLVER_WINS:
+                case RESOLVER_BCAST:
+                    if ( resolverType == ResolverType.RESOLVER_WINS && name.name != NbtAddress.MASTER_BROWSER_NAME && name.hexCode != 0x1d ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.